### PR TITLE
Fix a bug where gossip loops forever while splitting messages

### PR DESCRIPTION
**Pull request by sagar-solana**
#### Problem

Gossip tries to discard values that cannot fit in a single transmission.
The logic that does this fails to account for a border condition where a value is small enough by itself but doesn't leave enough room for the "vec" wrapper that needs to go around it. This causes the function to loop for ever

#### Summary of Changes

Fixed the check to see if a message can fit in a separate transmission

### CODE REVIEW

File: core/src/cluster_info.rs [hunk 1]
```diff
@@ -917,11 +917,14 @@ impl ClusterInfo {
         let mut messages = vec![];
         while !msgs.is_empty() {
```
Comment (aeyakovenko) [inline core/src/cluster_info.rs]: get rid of the while loop

File: core/src/cluster_info.rs [hunk 2]
```diff
@@ -917,11 +917,14 @@ impl ClusterInfo {
...
+            let base_size = serialized_size(&payload).expect("Couldn't check size");
+            let mut size = base_size;
             while let Some(msg) = msgs.pop() {
                 let msg_size = msg.size();
+                // If the message is too big to fit in this batch
                 if size + msg_size > MAX_PROTOCOL_PAYLOAD_SIZE as u64 {
-                    if msg_size < MAX_PROTOCOL_PAYLOAD_SIZE as u64 {
+                    // See if it can fit in the next batch
+                    if base_size + msg_size <= MAX_PROTOCOL_PAYLOAD_SIZE as u64 {
                         msgs.push(msg);
```
Comment (aeyakovenko) [inline core/src/cluster_info.rs]: this is the problem.  the code pushes back into the data structure it iterates over

Comment (sagar-solana) [reply core/src/cluster_info.rs]: but getting rid of the loop is the way to go.

Comment (sagar-solana) [reply core/src/cluster_info.rs]: @aeyakovenko how about now?

File: core/src/cluster_info.rs [hunk 3]
```diff
@@ -913,27 +913,35 @@ impl ClusterInfo {
...
-                        );
+        let mut payload = vec![];
+        let base_size = serialized_size(&payload).expect("Couldn't check size");
+        let mut size = base_size;
+        for msg in msgs {
+            let msg_size = msg.size();
+            // If the message is too big to fit in this batch
+            if size + msg_size > MAX_PROTOCOL_PAYLOAD_SIZE as u64 {
+                // See if it can fit in the next batch
+                if base_size + msg_size <= MAX_PROTOCOL_PAYLOAD_SIZE as u64 {
```
Comment (aeyakovenko) [inline core/src/cluster_info.rs]: aren't you counting base_size twice?

how about

max = MAX_PROTOCOL_PAYLOAD_SIZE - base_size;
size = 0;

File: core/src/cluster_info.rs [hunk 4]
```diff
@@ -913,27 +913,35 @@ impl ClusterInfo {
...
+                    if !payload.is_empty() {
+                        // flush the  current payload
+                        messages.push(payload);
                     }
-                    break;
+                    size = base_size + msg_size;
```
Comment (aeyakovenko) [inline core/src/cluster_info.rs]: why doesn't this just fall through to line 941?

File: core/src/cluster_info.rs (hunk 3 above)
Comment (sagar-solana) [reply core/src/cluster_info.rs]: Am I? Where? 
msg_size is the size of the crdsvalue itself.
Size is init to base size (which like you suggested I can move to rhs).
Size is then incremented by msg_size each time a new msg is added. 
When we move to a new transmission, size is reset to base size + rollover msg size and payload is init to carry the rollover message.

File: core/src/cluster_info.rs (hunk 4 above)
Comment (sagar-solana) [reply core/src/cluster_info.rs]: I think it can if I move base size over to the rhs in the of statements. Will do. Thanks!

Comment (sagar-solana) [reply core/src/cluster_info.rs]: I'll need to reset size to 0 and payload to a new vec, so I might as well init them to 941 would have done?

File: core/src/cluster_info.rs [hunk 5]
```diff
@@ -913,27 +913,37 @@ impl ClusterInfo {
...
+            let msg_size = msg.size();
+            // If the message is too big to fit in this batch
+            if payload_size + msg_size > max_payload_size as u64 {
+                // See if it can fit in the next batch
+                if msg_size <= max_payload_size as u64 {
+                    if !payload.is_empty() {
+                        // Flush the  current payload
+                        messages.push(payload);
+                        // Init the next payload
+                        payload = vec![msg];
```
Comment (aeyakovenko) [inline core/src/cluster_info.rs]: it looks like msg is encoded twice, here and after it falls through to the end of the loop.  get some sleep and write better unit tests :)

Comment (sagar-solana) [reply core/src/cluster_info.rs]: lol but it "continues" at the end of the if block?

Comment (sagar-solana) [reply core/src/cluster_info.rs]: @aeyakovenko did you see the `continue` at the end? it's not encoding anything twice, its just resetting state, adding the first new item and going next.

**Pull request merged**